Friendly Mod Export

dbent opened the pull request:
Fixes #1094

Allow CKAN to export the mod list in multiple formats.

- Plain Text
- Markdown
- BBCode
- CSV
- TSV

This is accomplished in the GUI by expanding the "Save as type" option. In the command line an extra option is added to the `list` command, `--export` that will output the list in the specified format (`text`, `markdown`, `bbcode`, `csv`, `tsv`) to stdout. Users can then redirect to a file as appropriate.

The exact format of the "pretty" formats is just what I came up with, suggestions welcome.

Also, I noticed that the `list` command has a `porcelain` option that's used in the [exact opposite sense that git uses that word](http://stackoverflow.com/questions/6976473/what-does-the-term-porcelain-mean-in-git). We should probably change that?

dbent commented:
This [gist](https://gist.github.com/dbent/9bcbb71cd50ca3b186ec) contains examples of all the formats.

Dazpoet commented:
I think the examples are awesome and really like them. I think that users might want a plain-text option which doesn't include our identifiers though but only `<mod name> <version>`. However I guess an argument can easily be made that supplying c/tsv makes making one of those on your own an extremely easy task.

pjf commented on Core/Exporters/IExporter.cs:
```diff
@@ -0,0 +1,17 @@
+namespace CKAN.Exporters
+{
+    /// <summary>
+    /// Represents an object that can 
```
Are you sure that? ;)

dbent replied:
... do whatever it puts its mind to because it's beautiful and special.

I'll fix that.

pjf commented on Cmdline/Action/List.cs:
```diff
@@ -31,66 +47,117 @@ public int RunCommand(CKAN.KSP ksp, object raw_options)
-                        else if (latest.version.IsEqualTo(current_version))
+                    if (current_version is ProvidesVersion)
+                    {
+                        // Skip virtuals for now.
```
Oh goodness, did we use to display these? Nice catch.

dbent replied:
That's just a whitespace change, add `?w=1` to a compare URL to force GitHub to [ignore whitespace in the diff](https://github.com/KSP-CKAN/CKAN/pull/1099/files?w=1).

pjf commented:
I haven't had a chance to test this, but:

- Adding a `CKAN.Exporters` namespace seems solid to me.
- All the code looks good.
- This will make lots of people happy.

In other words, if you're reading this and able to test this is working as intended, I'm totally happy for you to mash the 'merge' button, even if you're not a C# person. :)

pjf commented:
Also as discussed on IRC, git's use of porcelain is confusing. We're using it as "more machine readable", which is one of the ways in which git uses it. :)

dbent commented:
> I'm totally happy for you to mash the 'merge' button, even if you're not a C# person. :)

I'm *totes* a C# person, it's been my professional life for the last six years.

I have one more change I want to make before I mash the merge button: I don't like the duplicate `switch` statement for both the command line and GUI, so I'm going to make a single `IExporter` that takes a `ExportFileType` and performs the `switch` internally. Then both the GUI and command line will just need to know how to determine the `ExportFileType` and instantiate the meta-exporter.

dbent commented:
A few updates in the latest batch of changes:

  - The `IExporter` interface has been generalized from using a `TextWriter` to using a generic `Stream`. Now the same interface can be used for writing binary formats ([BSON](http://bsonspec.org/) piped over the network anyone?)
  - I implemented a meta `Exporter` class, which consolidates all the switching logic to choose the right `IExporter`. Now the GUI and command line simply instantiate an `Exporter` and call its `Export` method. Yay, for code reuse.
  - I modified the Markdown and BBCode formats to use bold instead of italics on the mod name, I personally think this looks nicer and is more legible.

dbent commented:
#rebuild

dbent commented:
I guess the `#rebuild` command doesn't work for Travis? Apparently closing and reopening should work...

dbent commented:
Interesting, closing and reopening the PR triggered a [new build](https://travis-ci.org/KSP-CKAN/CKAN/builds/66549612), but did not update the status of the PR, which is still linked to the old PR.

techman83 commented:
@dbent you can trigger a new build by adding commits, you just have to wait a little while. You can also go to the Travis page and re-run it as is, there is an action button in the top right of the box.

dbent commented:
Merging this since the changes since the review are minor.
